nbgrader/docs/source/user_guide/what_is_nbgrader.rst

```diff
@@ -0,0 +1,308 @@
+disadvantage...).  Because this is not a "captive system" that
+segments users from the data they are working on, you can't easily
+limit access to the underlying operating system or notebook file.
+However, if you design your course considering this, we hope you'll
+see that this is an advantage.
+
+
+
+Course directory
+----------------
+
+From the instructor side, all data is stored in files in the **course
+directory**.  Within this directory, files have the general structure
+``{nbgrader_step}/{student_id}/{assignment_id}/{notebook}.ipynb``.
+The ``{nbgrader_step}`` indicates the progress of the notebook:
+
+* ``source/`` contains the raw source notebooks made by the
+  instructor.  You run ``nbgrader generate_assignment`` to produce
+  the...
+* ``release/`` directory.  This contains a version with the "hidden
+  tests" removed and various cells marked read-only.  For both the
+  ``source`` and ``release`` steps, the ``student_id`` part of the
+  path is missing, since the notebooks are not personalized per
+  student.
+* The release version gets copied to the students, and copied back.
+  This can be through the "exchange" (below), or some other technique.
+* The ``submitted/`` step contains the students submitted versions.
+  By running ``nbgrader autograde``, the files are autograded (see
+  below) and copied to the...
+* ``autograded`` directory.  Autograding is described below, but
+  basically consists of executing the notebook and looking for failed
+  cells.  ``autograded`` stores the executed notebook.  One can also
+  use this for manual grading.  Running ``generate_feedback``, one gets...
+* ``feedback``.  This stores a ``.html`` (not ``.ipynb``) file with
+  the executed notebook, points, and comments from manual grading.
+  This can be returned to the students.
+
+
+But, there is a database in **gradebook.db** which stores students and
+grades.  Less obviously, it stores the assignments and the contents of
+hidden or immutable cells, such as the contents of the hidden tests
+or read-only cells.  This is used to restore these cells when students
+return them.
+
+
+nbgrader generate_assignment
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``generate_assignment`` converts the release notebook file from the
+source.  Like all of these commands, it is implemented as converters
+that implement various filters of the notebook source, in this case
+reading from a path in ``source/`` and writing to a path in
+``release/``.  Exact operations can be found within the
+``nbgrader.converters`` path.
+
+The main effect of ``generate_assignment`` is to strip all output
+cells and remove ``### BEGIN SOLUTION`` AND ``### BEGIN HIDDEN TEST``
+blocks, and record the contents of various read-only cells in the
+database.
+
+
+
+nbgrader autograde
+~~~~~~~~~~~~~~~~~~
+
+This is described below, but the core tasks are to re-insert the
+read-only cells (replace them with the known-good versions), and
+execute the entire notebook.
+
+
```

Comment:
In the UI, an instructor can, having ``autograded`` all submissions, manually grade them.

During this process, the instructor is shown each students submission, with a text-box under the ``answer`` cells [ref to come] where they can leave a comment, of "feedback"